### PR TITLE
Fix for Enable-AzureRmContextAutosave in Netcore (#5690).

### PR DESCRIPTION
**Pull request by twitchax**
## Checklist

- [x] I have read the [_Submitting Changes_](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#making-changes) section of [`CONTRIBUTING.md`](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md)
- [x] The title of the PR is clear and informative
- [x] The appropriate [change log has been updated](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#updating-the-change-log)
- [x] The PR does not introduce [breaking changes](https://github.com/Azure/azure-powershell/blob/preview/documentation/breaking-changes/breaking-changes-definition.md)
- [x] If applicable, the changes made in the PR have proper test coverage
- [ ] For public API changes to cmdlets:
    - [ ] the changes have gone through a [cmdlet design review](https://github.com/Azure/azure-powershell-cmdlet-review-pr)
    - [ ] the cmdlet markdown files were [generated using the `platyPS` module](https://github.com/Azure/azure-powershell/blob/preview/documentation/development-docs/help-generation.md)

### CODE REVIEW

**Comment by twitchax** [comment]
Ficing #5690.

**Comment by maddieclayton** [comment]
@twitchax Looks like a couple of profile tests are failing - can you take a look?

**Comment by twitchax** [comment]
@maddieclayton, fixed! :)

**Comment by maddieclayton** [comment]
on demand here: https://azuresdkci.westus2.cloudapp.azure.com/view/PowerShell/job/powershell-demand/591/ (and 592)

**Comment by twitchax** [comment]
@markcowl, can you let me know when you merge this?  I have another PR which depends upon this one.

File: src/Common/Commands.Common.Authentication/Authentication/ProtectedFileTokenCache.cs [hunk 1]
```diff
@@ -113,45 +96,70 @@ public override void Clear()
...
+                WriteCacheIntoFile();
+                // once the write operation took place, restore the HasStateChanged bit to false
+                HasStateChanged = false;
+            }
+        }
+
+        private void ReadFileIntoCache(string cacheFileName = null)
+        {
+            if(cacheFileName == null)
+                cacheFileName = ProtectedFileTokenCache.CacheFileName;
```
Comment (markcowl) [inline src/Common/Commands.Common.Authentication/Authentication/ProtectedFileTokenCache.cs]: use braces for the then block

File: src/Common/Commands.Common.Authentication/Authentication/ProtectedFileTokenCache.cs [hunk 2]
```diff
@@ -113,45 +96,70 @@ public override void Clear()
...
 
-        // Triggered right after ADAL accessed the cache.
-        void AfterAccessNotification(TokenCacheNotificationArgs args)
+        private void WriteCacheIntoFile(string cacheFileName = null)
         {
-            // if the access operation resulted in a cache update
-            EnsureStateSaved();
-        }
+            if(cacheFileName == null)
+                cacheFileName = ProtectedFileTokenCache.CacheFileName;
```
Comment (markcowl) [inline src/Common/Commands.Common.Authentication/Authentication/ProtectedFileTokenCache.cs]: braces for then block

File: src/Common/Commands.Common.Authentication/Authentication/ProtectedFileTokenCache.cs [hunk 3]
```diff
@@ -113,45 +96,70 @@ public override void Clear()
...
+        // Triggered right after ADAL accessed the cache.
+        void AfterAccessNotification(TokenCacheNotificationArgs args)
+        {
+            // if the access operation resulted in a cache update
+            EnsureStateSaved();
+        }
+
+        void EnsureStateSaved()
+        {
+            if (HasStateChanged)
```
Comment (markcowl) [inline src/Common/Commands.Common.Authentication/Authentication/ProtectedFileTokenCache.cs]: this isn't quite the same logic as before - we only want to write the cache once after a state change, so we want to write and clear the dirty bit in an atomic operation and keep other threads from doing so.

We can do this with a pattern like:

```C#
if (HasStateChanged)
{
   WriteCacheInfoFile()
}
```

and inside WriteCacheInfoFile:

``` C#
lock(lockFile)
{
  if (HasStateChanged)
  {  
   store.WriteFile(...);
   HasStateChanged =  false;
  }
}
```

or an equivalent.

File: src/Common/Commands.Common.Authentication/Authentication/ProtectedFileTokenCache.cs [hunk 4]
```diff
@@ -73,27 +73,10 @@ public ProtectedFileTokenCache(string cacheFile, IDataStore store = null)
...
-                            _store.DeleteFile(fileName);
-                        }
-                    }
-                }
-                
-                // Create the file to start with
-                _store.WriteFile(CacheFileName, ProtectedData.Protect(Serialize(), null, DataProtectionScope.CurrentUser));
-            }
+            ReadFileIntoCache(fileName);
+
```
Comment (markcowl) [inline src/Common/Commands.Common.Authentication/Authentication/ProtectedFileTokenCache.cs]: HasStateChanged=true;

Comment (twitchax) [reply src/Common/Commands.Common.Authentication/Authentication/ProtectedFileTokenCache.cs]: Nice catch on this one.  Gave comment to your requested changes. :)

**Comment by twitchax** [comment]
@markcowl, addressed feedback.

**Comment by markcowl** [comment]
@twitchax @maddieclayton @praries880 @cormacpayne Waiting until after code complete to merge

**Comment by twitchax** [comment]
@markcowl, can we merge now that the release branch has been cut?

**Comment by twitchax** [comment]
@markcowl, @maddieclayton, can we merge now that the release branch is cut?

**Comment by maddieclayton** [comment]
@markcowl mentioned something about merging this PR into one of his own

**Comment by twitchax** [comment]
@markcowl, when do you plan on doing this?  I have another PR which depends on this one.

**Comment by markcowl** [comment]
on demand run here: https://azuresdkci.westus2.cloudapp.azure.com/view/PowerShell/job/powershell-demand/653/

**Pull request merged**